### ISSUE TITLE
Feature: Import CSV/TSV files

### ISSUE THREAD

**Issue by david-allison** [ankidroid/Anki-Android#6772]
Anki Desktop allows the import of CSV/TSV, and makes use of the auto-detection of CSV files which is built into Python.

Splintering off this task from #3517 as this is close to complete locally.

There will likely be follow up support tickets which would best be associated with a specific ticket rather than an 'Epic'. This is a significant port of Python and C code, and there is a large likelihood for bugs.

**Comment by david-allison-1** [comment]
**Planned UI Design [DRAFT]**

**Current Anki UI**

![image](https://user-images.githubusercontent.com/62114487/88990102-0dd2b080-d2d5-11ea-8398-e5ae484428d9.png)


**AnkiDroid**

* Wizard with 3 screens: **"File"**, **"Map & Preview"**, **"Import"**
* **File**
  * Explanation, "load file" button and "Help" button
  * Responsible for file selection, basic parsing, and pre-warning about unsolvable issues.
     * Empty first field
     * Field appears twice in input
     * Incorrect Field Counts
  * Errors will be displayed to the user, along with the line number where they occurred. User will have a chance to refresh if they change the file, or select another file.
  * Line numbers in errors must ensure to map to actual line numbers (for example, comments, and the header of the CSV are ignored).
  * Users will be informed that any issues will be ignored if they continue.
* **Map & Preview**
  * Mapping the note type is the most important thing. This should be first, and likely done in a fragment
      * We should highlight good note types (with more or equal amounts of fields)
      * The number of fields should be displayed to the user when selecting a Note Type
      * It should be recommended to create a new Note Type if there are no suitable options
          * [V2?] Add transition to "Manage Note Types"/Automate the adding of a New Note Type
    * Once a note type is selected, a row should be read from the CSV and used to allow the user to make informed decisions for the field to map to the fields of the note type
       * **TODO:** UI to perform mapping?
       * As in Anki, "Map to [field]" may only be set once per note-type field.
  * Back button on this screen should prompt before leaving
  * **TODO** - other controls on Mapping
    * *Fields separated by*
      * The sniffing algorithm is typically very good - low priority
   * TODO: Warn users about first field conflicts before they import
  * Preview - Self explanatory - Preview Window allowing user to skip through cards to see an active preview.
* **Import**

**Comment by david-allison-1** [comment]
Rough functional plan:

```ocaml
Inputs: Path, SizeCheckConfirmation

LoadPath = Path -> File | IOException
ReloadFile = File -> File | IOException
FileSizeStatus = Ok | TooLarge of bytes: long
SizeWarning = File -> FileSizeStatus
SizeCheckConfirmation = FileSizeConfirmation -> Ok | Cancel
SizeCheck = File -> SizeCheckConfirmation -> Result<File, UserCancelled>
EncodingIssue = File -> Result<File, CSVEncodingException>
ReadError = IOException | UserCancelled | CSVEncodingException
ParseAsCsv = File -> SizeWarning >> EncodingIssue -> CsvFile | ReadError
ObtainWarnings = CsvFile -> [Warning]
ValidCsv = { fieldCount : int ; lines: [Field[fieldCount]] ;  tags: Option<[String]> }
AcknowledgeWarnings = CsvFile >> ObtainWarnings -> ValidCsv | UserCancelled

(* We ignore the field separator character for now 
Python autodetection should be sufficient and it increases complexity *)
----

GetRecommendedNoteTypes = ValidCsv -> [NoteType]
FieldMappingResult = FieldName | Tags | None
FieldMapping = { field:index ; mappedTo: FieldMappingResult }
FieldMappings = { FieldMapping[ValidCsv.fieldCount]; NoteType }
SelectNoteType = ValidCsv -> NoteType
SelectFieldMappings = ValidCsv -> NoteType -> FieldMappings
Preview = ValidCsv -> line -> FieldMapping -> IncludeHtml -> Option<Deck> -> CardPreview

ImportType = Update of [Tags] | Ignore | Add

ImportData = { ValidCsv; FieldMappingResult; IncludeHtml; ImportType; NoteType; }

Warning = Conflict | CardNotGenerated
DuplicateCheck = ImportData -> Warning[]

----

Import = ImportData -> Completed

```

**Comment by david-allison-1** [comment]
Needs a GUI - open for contributions, I'll be spending the next while getting the Rust conversion through, so won't have the time to get this through.

**Comment by VamsiKrishnaCommits** [comment]
Can I work on this? this seems to be a useful feature @david-allison-1

**Comment by david-allison-1** [comment]
> Can I work on this? this seems to be a useful feature @david-allison-1

This was part of a GSOC project - https://github.com/ankidroid/Anki-Android/wiki/GSOC-proposal-2021#csv-import

Let me check with the other maintainers - ideally it's an enthusiastic yes, but I'd want to make sure it's okay, as it may get in the way of other proposals.

**Comment by david-allison-1** [comment]
@VamsiKrishnaCommits - Result of the discussion: absolutely! I'll note this on the GSoC page

Thanks!

**Comment by almas33** [comment]
Hi @VamsiKrishnaCommits are you working on this
@david-allison-1 this idea is struck in GSoC idea list , If this is still available I would like to submit proposal on this

**Comment by lamvankhoat1** [comment]
Waited a long time for this function, but it hasn't been released yet 🤣

**Comment by david-allison-1** [comment]
Just needs a GUI! 

it's not on my radar right now as I'll spend most of 2.16 catching up with Anki Desktop functionality, but it's open to contributions

**Comment by lamvankhoat1** [comment]
it's been so long, i'm still waiting

**Comment by mikehardy** [comment]
Me too! Eagerly awaiting anyone that wants to contribute their time to develop the feature. Patience is a virtue :lotus_position: :laughing:

**Comment by david-allison-1** [comment]
Guess that'll be me.

Working a functional prototype using the Anki Desktop UI, then will improve UX via suggestions in my above comments

<img height="300px" src="https://user-images.githubusercontent.com/62114487/130033005-f8c19c20-83f5-42d3-8eab-573f00429d49.png" >

**Comment by lamvankhoat1** [comment]
> Guess that'll be me.
> 
> Working on a functional prototype using the Anki Desktop UI, then will improve UX via suggestions in my above comments
> 
> <img alt="" height="300px" src="https://user-images.githubusercontent.com/62114487/130033005-f8c19c20-83f5-42d3-8eab-573f00429d49.png">

It's amazing, the most attractive function of Ankidroid

**Comment by lamvankhoat1** [comment]
May I ask this, will this feature come out in this version 2.16?

**Comment by mikehardy** [comment]
Most likely not - 2.16 already has the most incredible backlog of features and fixes and changes going in, and putting anything else in it would be a bad idea from a release management / stability perspective.
2.16 will go as soon as #9249 scoped storage is done, to get all the current work out

**Comment by Arthur-Milchior** [comment]
> Guess that'll be me.
> 
> Working on a functional prototype using the Anki Desktop UI, then will improve UX via suggestions in my above comments
> 
> <img alt="" height="300px" src="https://user-images.githubusercontent.com/62114487/130033005-f8c19c20-83f5-42d3-8eab-573f00429d49.png">

Does this image comes from an implementation you actually did? I don't see link to any related code in this thread.

**Comment by david-allison** [comment]
> Does this image comes from an implementation you actually did? I don't see link to any related code in this thread.

https://github.com/ankidroid/Anki-Android/pull/9437 is the latest code

**Comment by dae** [comment]
Just a heads up that @rumovz's work on a new CSV importer/exporter was merged into Anki's main branch yesterday. The actual importing/exporting is handled in Rust, and there is a web UI for the field mapping. It includes some new functionality like per-row notetypes that the old Python implementation did not have.

While AnkiDroid won't be able to take advantage of it until AnkiDroid's Rust backend migration work is further along, I thought I'd let you know in case you wished to redirect efforts on this separate implementation to something else.

**Comment by Arthur-Milchior** [comment]
Thank you extremely much @dae for the information. Thanks also @
FYI, AnkiDroid accepted 4 Google Summer of Code mentees this year. One being CSV import. Good news is that the mentee, @BrayanDSO,  also planned to do other work, so the other task can be prioritized. But clairly it requires to adapt the project.

I just did a `git pull` and `./run` on anki's main, and still see the importer. Is there anything special to do to see the web UI? 

Also, do you know whether the back-end API is stable or do you intend for it to potentially change soon. Because, as @krmanik states on discord, we can make our UI so that it works with your back-end, having a common interface. But ideally, we'd do so if we expect it not to change soon

**Comment by dae** [comment]
The new code can be enabled in the first tab of the preferences screen.

The backend interface is strongly typed, so if it does happen to change in the future, you'd find out at compile time. But I'd recommend you consider making use of the Svelte frontend as well, as that way you get the interface mostly for free, instead of having to implement all the UI logic like field mapping. This is a common pattern in the desktop code - other screens like the deck options, card info, notetype changing and so on also have both backend and frontend components that are intended to be used across all clients.

**Comment by krmanik** [comment]
The UI using Svelte can be used but we have to consider following 
1. Responsive so it can adapt different screen size, css can be used
2. Localisation, it is already provided but some new language may not added
3. It will not be native so may be we get issues for it
4. It needs to load in webview along with implementation of localhost to serve file (because loading from `file:///` may give errors)

But the Svelte UI and rust backend will be true cross platform implementation. 

I have tried to implement for stats but didn't understand what should be next step.
https://github.com/ankidroid/Anki-Android/issues/10751#issuecomment-1114994683
>I have tried this for generating `graphs.html`, `graphs.js` and `graphs.css`. I have clone the anki repository and run the bazel build in web directory in [qt/aqt/data/web](https://github.com/ankitects/anki/tree/main/qt/aqt/data/web) and [ts](https://github.com/ankitects/anki/tree/main/ts). I get the generated `.js` and `.html` file in .bazel/bin folder.
>
>But how the data related to stats will be sent to graphs page when request sent from the graphs page? I think JavaScript interface will be used or local http server like [nanohttpd](https://github.com/NanoHttpd/nanohttpd) will be used to get stats data from java/kotlin class.

**Comment by dae** [comment]
1-2 should already be handled, and these pages are working in AnkiMobile. If there are Android-specific issues, patches are welcome :-)

The frontend sends and receives binary protobuf payloads in POST requests. The desktop uses a built-in webserver, and AnkiMobile uses a custom URL scheme. You may be able to use the latter approach in AnkiDroid as well, by intercepting requests for ankischeme://something with https://developer.android.com/reference/android/webkit/WebViewClient#shouldInterceptRequest(android.webkit.WebView,%20android.webkit.WebResourceRequest), and returning the relevant data. This might be the contents of the requested .html/.js file, or for dynamic requests, it would take the received data, feed it to the relevant backend function, and then return the resulting blob in the response body.

**Comment by mikehardy** [comment]
That does sound right on and perhaps a lot easier than on pure technical merit vs some codegen/JSI interface and definitely easier ecosystem-wise if we do things the same way AnkiMobile does (custom scheme / protobuf posting), thanks @dae

**Comment by krmanik** [comment]
[android.webkit.WebResourceRequest)](https://developer.android.com/reference/android/webkit/WebViewClient#shouldInterceptRequest(android.webkit.WebView,%20android.webkit.WebResourceRequest)) is easier implementation than other options. Thanks

**Comment by BrayanDSO** [comment]
Hi. As upstream was implementing CSV importer very recently, I wanted to wait until it was done so I could assess it better

I think it's finished now, so I can start it.
I like more doing the Svelte approach because this way I wouldn't do the temporary job of implementing a already deprecated feature
I'll be somewhat busy personally until July 14th so my plan was to finish almost all of the preferences' project until then, since I have figured how to do everything on it
After it, I'll assess the new approach of the CSV importer properly

From discord (https://discord.com/channels/368267295601983490/977324845811531826/991263309749833758)

**Comment by mikehardy** [comment]
Great yes! I'm happy we've got the backend coming up to date with all the work there so this is possible, if you're comfortable doing it this way. Hopefully #11644 is in today or tomorrow

**Comment by dae** [comment]
When it comes to bundling the js/html files, my thinking was that this is better done in https://github.com/ankidroid/Anki-Android-Backend/, so that the included files are guaranteed to match the commit used to build the backend. I made an initial attempt at it, but had to roll it back since I didn't have time to try to figure out how to integrate it correctly into the build: https://github.com/ankidroid/Anki-Android-Backend/commit/c6f8ee126866c2d4d5b97244a11a05f5568bb46f

Another advantage of doing it that way is that it also generates the buildinfo.txt file that is needed for building the backend.

I did not test it on the AnkiDroid side though, so I don't know how those assets bundled into the -backend aar file would be readable on AnkiDroid's end.

**Comment by dae** [comment]
Should be unblocked when https://github.com/ankidroid/Anki-Android-Backend/pull/217 is merged and AD points to a new release.

**Comment by BrayanDSO** [comment]
With the new backend and based on https://github.com/krmanik/Anki-Android/pull/23, I got it to work here locally. Thanks to Mani and Damien for it.

Currently, I've got only to:
1. implement a progress dialog to be updated with the backend progress (there is a Progress class in the backend which I think I can use) 
2. Figure out a good way to organize the code
3. Maybe see if I can inject/modify the page CSS, so it will match AnkiDroid themes' colors

---

The only problems I've encountered so far are a white screen while loading the page and the fields flickering on configuration changed. Found them on the desktop version as well (Anki 2.1.54 Qt6, Windows 10). 

## White flashing and flickering on AnkiDroid


https://user-images.githubusercontent.com/69634269/183265954-a01488ec-3919-4fc9-b487-1f68bfeaadf7.mp4

## White flashing on desktop
https://user-images.githubusercontent.com/69634269/183265398-36d25db6-0687-4ac7-a261-3e8fb28e2dab.mp4

## Flickering on desktop
https://user-images.githubusercontent.com/69634269/183265366-8852cc4e-7cd4-41a0-ab00-7328359e9a4f.mp4

---

@dae, are these known issues on the desktop version?

I've found some "white loading" issues there (e.g. https://github.com/ankitects/anki/issues/1910), do you know if it affects only the web/html screens? 

Right now I don't know how to fix/workaround it, so I'll leave it. If I get time until GSoC's final deadline, I'll try looking into the flickering problem.

**Comment by dae** [comment]
@Rumovz has fixed the glitching when changing notetypes. The flash on page load happens because it takes a while to parse the JS and load the data from a network request. You could probably work around it by hiding the webview to start with (eg opacity 0), and then revealing it once the page has loaded (eg by awaiting anki.setupImportCsvPage() then sending a message back to Kotlin).

**Comment by dae** [comment]
- Colours are mostly controlled by CSS vars, which you can adjust at runtime. See _vars.scss
- If/when #11849 lands, you can use withProgress(): https://github.com/ankidroid/Anki-Android/blob/7c917defa789bf1106d34408a30509349fd430df/AnkiDroid/src/main/java/com/ichi2/anki/BackendImporting.kt#L29

**Comment by BrayanDSO** [comment]
I've managed to inject the CSS and show the progress dialog. Thank you very much, Damien.

I've configured the webview to only show after the page is loaded. It reduced the time the screen is white, but the white flash still appears.

I've tested on other Anki HTML screens (more below) and the white flashing doesn't happen on the "Statistics" screen while it flashes in gray on the "Card info" screen (and for some reason it flashes in black on Anki desktop (on dark mode)), so I guess it is some sort of problem with the JS.

![2022-08-08_13-18-27](https://user-images.githubusercontent.com/69634269/183473542-c5a3bb1f-4be5-4d8c-9444-2d3519a822de.png)
![2022-08-08_13-29-27](https://user-images.githubusercontent.com/69634269/183473546-7d6250a1-221f-4c5b-80d0-797e73ef234e.png)

Only pending things for myself are
- polish the code
- Fix the position of "Field separator" and "Allow HTML in fields" dropdown menu/toggle when there are many fields
    - ![image](https://user-images.githubusercontent.com/69634269/183474787-89796ec9-0bbb-4555-a8c6-382146e8211a.png)
- Be able to scroll horinzontally the fields table when it is too big to fit the screen (as the example above)

---

Probably unrelated, but while I was at it, I've tried to use the "Card info" and "Statistics" screens as well. It was very easy to setup them and both work out of the box, which is amazing!

Found some issues for the implementator (probably myself if nobody does it before me) solve in the feature

# Card info

CardInfo shows "(No card to display)" for a moment before loading the screen

![image](https://user-images.githubusercontent.com/69634269/183472821-18ae782b-da0a-47f2-b192-511afb276cf8.png)

https://user-images.githubusercontent.com/69634269/183472636-fefe620a-26ae-46a8-8f82-b3306b8e9740.mp4

Same happens on Anki desktop (2.1.54, Qt6, Windows 10), so may be an upstream's issue (on the video, it can be seen on the second time I open Card Info, probably only visible frame-by-frame)

https://user-images.githubusercontent.com/69634269/183472556-79619d0a-0f21-4722-bd0e-d32c2545a412.mp4

![image](https://user-images.githubusercontent.com/69634269/183473095-5c8f5dc9-c51b-416b-bcea-4ce20078c276.png)

# Statistics

Only aligning the texts with the radio buttons to do. Should be easy to do it on the CSS

![image](https://user-images.githubusercontent.com/69634269/183473235-4a71588d-1b6d-4581-9e9e-e9aeec811d50.png)

![image](https://user-images.githubusercontent.com/69634269/183473276-5b71a827-6d54-4cdd-ae24-cf85dadb6385.png)

**Comment by dae** [comment]
> I've tested on other Anki HTML screens (more below) and the white flashing doesn't happen on the "Statistics" screen while it flashes in gray on the "Card info" screen (and for some reason it flashes in black on Anki desktop (on dark mode)), so I guess it is some sort of problem with the JS.

Could be - PRs in the upstream repo most welcome :-)

> Fix the position of "Field separator" and "Allow HTML in fields" dropdown menu/toggle when there are many fields

This displays correctly when importing on the desktop and shrinking the window width down, and it displays correctly on iOS too - the table can be scrolled with a finger and the menu/toggles don't go offscreen. So this might be an Android-specific issue, where `        <meta name="viewport" id="viewport" content="width=device-width" />
` is behaving differently than on iOS. Any ideas @krmanik?

**Comment by BrayanDSO** [comment]
Nevermind, on my phone it scrolls/fit the screen perfectly without any changes on the code.

I was initially testing on the emulator (SDK 25, Pixel 2, 1920x1080), where I found the problem. I'll see if I sort out if it's a problem with older webview versions, older Android versions or DPI specific.

**Comment by dae** [comment]
How does webview updating work on Android? On iOS the webview version is bound to the iOS release, but I thought Android updated independently. The current minimum target version for the HTML/JS is iOS 13/Chromium 77, so it may fail or render incorrectly on older Chrome versions.

**Comment by BrayanDSO** [comment]
> How does webview updating work on Android? On iOS the webview version is bound to the iOS release, but I thought Android updated independently.

We can update it on Google Play. AFAIK, our minSDK (21) has support for Chromium 77+.

I can't test it currently because the one that can be downloaded on Android Studio is shipped with an older SDK. But as I've fixed the problem on other versions by updating the webView, it should work there too. Anyway, I'll leave that for the alpha/beta testers to confirm.

**Comment by krmanik** [comment]
I think the issues is resolved now. It is related to webview.

_Currently I am away from the system which have Android Studio/WSL, so I may not check it_
